content/docs/concepts/remote_write_spec_2_0.md

```diff
@@ -0,0 +1,394 @@
+- MUST have label names sorted in lexicographical order.
+- MUST NOT contain any empty label names or values.
+
+Sender MUST only send valid metric names, label names, and label values:
+
+<!---
+TODO(bwplotka): Add mention for UTF-8 support going forward, see https://docs.google.com/document/d/1PljkX3YLLT-4f7MqrLt7XCVPG3IsjRREzYrUzBxCPV0/edit?disco=AAAA4gSqQ7g
+-->
+- Metric names MUST adhere to the regex `[a-zA-Z_:]([a-zA-Z0-9_:])*`.
+- Label names MUST adhere to the regex `[a-zA-Z_]([a-zA-Z0-9_])*`.
+- Label values MAY be any sequence of UTF-8 characters .
+
+Receiver MAY impose limits on the number and length of labels, but this will be receiver-specific and is out of scope for this document.
+
+Label names beginning with "__" are RESERVED for system usage and SHOULD NOT be used, see [Prometheus Data Model](https://prometheus.io/docs/concepts/data_model/).
+
+#### Samples
+
+<!---
+TODO(bwplotka): Does this apply to histograms?
+-->
+Sender MUST send samples for any given TimeSeries in timestamp order. Sender MAY send multiple requests for different series in parallel.
+
+Sender MUST send stale markers when a time series will no longer be appended to.
+
+Stale markers MUST be signalled by the special NaN value `0x7ff0000000000002`. This value MUST NOT be used otherwise.
+
+Typically, Sender can detect when a time series will no longer be appended to using the following techniques:
+
+1. Detecting, using service discovery, that the target exposing the series has gone away
+1. Noticing the target is no longer exposing the time series between successive scrapes
+1. Failing to scrape the target that originally exposed a time series
+1. Tracking configuration and evaluation for recording and alerting rules
+
+#### Metadata
+
+Metadata SHOULD follow the official guidelines for [TYPE](https://prometheus.io/docs/instrumenting/writing_exporters/#types) and [HELP](https://prometheus.io/docs/instrumenting/writing_exporters/#help-strings).
+
+#### Exemplars
+
+<!---
+TODO(bwplotka): Anything to say here?
+-->
+TBD
+
+#### Created Timestamp
+
+<!---
+TODO(bwplotka): TBD
+-->
+TBD
+
+### Responses
+
+Receiver ingesting all samples successfully MUST return HTTP 200 status code. In such a successful case, the response body from the Receiver SHOULD be empty; Sender MUST ignore the response body. The response body is RESERVED for future use.
+
+The following subsections specify Sender and Receiver semantics around write errors.
+
+#### Partial Write
+
+Sender SHOULD use Prometheus Remote Write to request write of multiple samples, across multiple series. As a result, Receiver MAY ingest valid samples within a write request that contains invalid or otherwise unwritten samples, which represents a partial write case.
+
+In a partial write case, Receiver MUST NOT return HTTP 200 status code. Receiver MUST provide a human-readable error message in the response body. Sender MUST NOT try and interpret the error message, and SHOULD log it as is.
```

Comment:
> Sender MUST NOT try and interpret the error message, and SHOULD log it as is.

Can we standardize a way for senders to interpret the error message? 

Today while debugging failed sample errors with Remote-Write 1.0, I was wishing I could visualize (from the sender's perspective) the types of errors being returned from the server and the number of samples discarded per error type the server generated. 

I don't think this information has to be in the response body, but it could be a header field or another place where machines can read it.

Reply:
Thanks! 

We discussed this with @cstyan a lot and decided to skip the standirazation of error message for now. As mentioned on Slack our recommendation would be organically grow this feature - so propose it to Prometheus (both sender and receiver) and see how it will be actually used/implemented. Then we can judge it and decide if we need this in 2.x or 3.0. This feature feels useful, but we might need data if "programmatic" use will be there. There is some more adoption work and rabbit holes (what error types etc?) to discuss if we ever propose it, so more data first would be nice. So far it feels error log should suffice. 

We extended the sentence to:

```
... Receiver MUST provide a human-readable error message in the response body. The Receiver's error SHOULD contain information about the amount of the samples being rejected and for what reasons.
```

Hope that's ok! And thanks for amazing idea!